File: Xero.WebApi/Data/NrtRunDefinitionRepository.cs
using System.Text.Json;
using Dapper;
using Xero.DataAcquisition;
using Xero.SmartComparer;
using Xero.WebApi.Models;

namespace Xero.WebApi.Data;

/// <summary>
/// CRUD operations for <c>nrt_run_definitions</c>.
/// Uses Dapper + JSONB for nested config objects.
/// </summary>
public sealed class NrtRunDefinitionRepository
{
    private readonly IDbConnectionFactory _factory;
    private readonly string               _connectionString;

    private static readonly JsonSerializerOptions JsonOpts =
        new() { PropertyNameCaseInsensitive = true };

    public NrtRunDefinitionRepository(IDbConnectionFactory factory, string connectionString)
    {
        _factory          = factory;
        _connectionString = connectionString;
    }

    // ── Queries ───────────────────────────────────────────────────────────────

    public async Task<IReadOnlyList<NrtRunDefinitionSummary>> GetAllAsync(CancellationToken ct)
    {
        using var conn = _factory.CreateConnection(_connectionString);
        var rows = await conn.QueryAsync<NrtRunDefinitionSummary>(new CommandDefinition(
            @"SELECT definition_id      AS DefinitionId,
                     name               AS Name,
                     description        AS Description,
                     scope              AS Scope,
                     scenario_name      AS ScenarioName,
                     reference_version  AS ReferenceVersion,
                     target_version     AS TargetVersion,
                     ref_commandline    AS RefCommandLine,
                     target_commandline AS TargetCommandLine,
                     created_at         AS CreatedAt,
                     updated_at         AS UpdatedAt,
                     is_active          AS IsActive
              FROM   nrt_run_definitions
              WHERE  is_active = true
              ORDER  BY name",
            cancellationToken: ct));
        return rows.AsList();
    }

    public async Task<NrtRunDefinitionDto?> GetByIdAsync(Guid id, CancellationToken ct)
    {
        using var conn = _factory.CreateConnection(_connectionString);
        var row = await conn.QuerySingleOrDefaultAsync<RawDefinitionRow>(new CommandDefinition(
            @"SELECT definition_id      AS DefinitionId,
                     name               AS Name,
                     description        AS Description,
                     scope              AS Scope,
                     scenario_name      AS ScenarioName,
                     reference_version  AS ReferenceVersion,
                     target_version     AS TargetVersion,
                     ref_commandline    AS RefCommandLine,
                     target_commandline AS TargetCommandLine,
                     created_at         AS CreatedAt,
                     updated_at         AS UpdatedAt,
                     is_active          AS IsActive,
                     reference_db::text AS ReferenceDbJson,
                     target_db::text    AS TargetDbJson,
                     compare_settings::text AS CompareSettingsJson,
                     output_settings::text  AS OutputSettingsJson,
                     column_schema::text    AS ColumnSchemaJson
              FROM   nrt_run_definitions
              WHERE  definition_id = @Id",
            new { Id = id },
            cancellationToken: ct));

        if (row is null) return null;

        return new NrtRunDefinitionDto
        {
            DefinitionId      = row.DefinitionId,
            Name              = row.Name,
            Description       = row.Description,
            Scope             = row.Scope,
            ScenarioName      = row.ScenarioName,
            ReferenceVersion  = row.ReferenceVersion,
            TargetVersion     = row.TargetVersion,
            RefCommandLine    = row.RefCommandLine,
            TargetCommandLine = row.TargetCommandLine,
            CreatedAt         = row.CreatedAt,
            UpdatedAt         = row.UpdatedAt,
            IsActive          = row.IsActive,
            Reference    = Deserialize<DbSettingsDto>(row.ReferenceDbJson)         ?? new DbSettingsDto(),
            Target       = Deserialize<DbSettingsDto>(row.TargetDbJson)            ?? new DbSettingsDto(),
            Compare      = Deserialize<CompareSettingsDto>(row.CompareSettingsJson) ?? new CompareSettingsDto(),
            Output       = Deserialize<OutputSettingsDto>(row.OutputSettingsJson)   ?? new OutputSettingsDto(),
            ColumnSchema = Deserialize<ColumnDef[]>(row.ColumnSchemaJson)           ?? Array.Empty<ColumnDef>(),
        };
    }

    private sealed class RawDefinitionRow
    {
        public Guid      DefinitionId      { get; init; }
        public string    Name              { get; init; } = "";
        public string?   Description       { get; init; }
        public string?   Scope             { get; init; }
        public string    ScenarioName      { get; init; } = "";
        public string    ReferenceVersion  { get; init; } = "";
        public string    TargetVersion     { get; init; } = "";
        public string?   RefCommandLine    { get; init; }
        public string?   TargetCommandLine { get; init; }
        public DateTime  CreatedAt         { get; init; }
        public DateTime  UpdatedAt         { get; init; }
        public bool      IsActive          { get; init; }
        public string?   ReferenceDbJson   { get; init; }
        public string?   TargetDbJson      { get; init; }
        public string?   CompareSettingsJson { get; init; }
        public string?   OutputSettingsJson  { get; init; }
        public string?   ColumnSchemaJson    { get; init; }
    }

    // ── Mutations ─────────────────────────────────────────────────────────────

    public async Task<Guid> CreateAsync(SaveNrtRunDefinitionRequest req, CancellationToken ct)
    {
        using var conn = _factory.CreateConnection(_connectionString);
        return await conn.ExecuteScalarAsync<Guid>(new CommandDefinition(
            @"INSERT INTO nrt_run_definitions
                  (name, description, scope, scenario_name, reference_version, target_version,
                   ref_commandline, target_commandline,
                   reference_db, target_db, compare_settings, output_settings, column_schema)
              VALUES
                  (@Name, @Description, @Scope, @ScenarioName, @ReferenceVersion, @TargetVersion,
                   @RefCommandLine, @TargetCommandLine,
                   @ReferenceDb::jsonb, @TargetDb::jsonb,
                   @CompareSettings::jsonb, @OutputSettings::jsonb, @ColumnSchema::jsonb)
              RETURNING definition_id",
            new
            {
                req.Name,
                req.Description,
                req.Scope,
                req.ScenarioName,
                req.ReferenceVersion,
                req.TargetVersion,
                req.RefCommandLine,
                req.TargetCommandLine,
                ReferenceDb     = Serialize(req.Reference),
                TargetDb        = Serialize(req.Target),
                CompareSettings = Serialize(req.Compare),
                OutputSettings  = Serialize(req.Output),
                ColumnSchema    = Serialize(req.ColumnSchema),
            },
            cancellationToken: ct));
    }

    public async Task<bool> UpdateAsync(Guid id, SaveNrtRunDefinitionRequest req, CancellationToken ct)
    {
        using var conn = _factory.CreateConnection(_connectionString);
        var affected = await conn.ExecuteAsync(new CommandDefinition(
            @"UPDATE nrt_run_definitions
              SET name               = @Name,
                  description        = @Description,
                  scope              = @Scope,
                  scenario_name      = @ScenarioName,
                  reference_version  = @ReferenceVersion,
                  target_version     = @TargetVersion,
                  ref_commandline    = @RefCommandLine,
                  target_commandline = @TargetCommandLine,
                  reference_db       = @ReferenceDb::jsonb,
                  target_db          = @TargetDb::jsonb,
                  compare_settings   = @CompareSettings::jsonb,
                  output_settings    = @OutputSettings::jsonb,
                  column_schema      = @ColumnSchema::jsonb
              WHERE definition_id = @Id AND is_active = true",
            new
            {
                Id               = id,
                req.Name,
                req.Description,
                req.Scope,
                req.ScenarioName,
                req.ReferenceVersion,
                req.TargetVersion,
                req.RefCommandLine,
                req.TargetCommandLine,
                ReferenceDb     = Serialize(req.Reference),
                TargetDb        = Serialize(req.Target),
                CompareSettings = Serialize(req.Compare),
                OutputSettings  = Serialize(req.Output),
                ColumnSchema    = Serialize(req.ColumnSchema),
            },
            cancellationToken: ct));
        return affected > 0;
    }

    /// <summary>Soft-deletes by setting <c>is_active = false</c>.</summary>
    public async Task<bool> SoftDeleteAsync(Guid id, CancellationToken ct)
    {
        using var conn = _factory.CreateConnection(_connectionString);
        var affected = await conn.ExecuteAsync(new CommandDefinition(
            @"UPDATE nrt_run_definitions
              SET is_active = false
              WHERE definition_id = @Id AND is_active = true",
            new { Id = id },
            cancellationToken: ct));
        return affected > 0;
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private static string Serialize<T>(T obj) => JsonSerializer.Serialize(obj);

    private static T? Deserialize<T>(string? json)
    {
        if (string.IsNullOrWhiteSpace(json)) return default;
        try { return JsonSerializer.Deserialize<T>(json, JsonOpts); }
        catch { return default; }
    }
}

File: Xero.JavaWebApi/src/main/java/com/xero/webapi/repository/NrtRunDefinitionRepository.java
package com.xero.webapi.repository;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.xero.webapi.model.ColumnDef;
import com.xero.webapi.model.CompareSettingsDto;
import com.xero.webapi.model.DbSettingsDto;
import com.xero.webapi.model.NrtRunDefinitionDto;
import com.xero.webapi.model.NrtRunDefinitionSummary;
import com.xero.webapi.model.OutputSettingsDto;
import com.xero.webapi.model.SaveNrtRunDefinitionRequest;
import io.github.resilience4j.retry.annotation.Retry;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.dao.EmptyResultDataAccessException;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.stereotype.Repository;

import java.sql.Timestamp;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.List;
import java.util.UUID;

/**
 * CRUD operations for {@code nrt_run_definitions}. Nested config objects are stored
 * as JSONB and serialised via the standalone {@code jsonbObjectMapper} bean.
 *
 * <p>All methods are wrapped with a Resilience4j {@code @Retry("database")} so transient
 * connectivity / timeout failures are recovered automatically.
 */
@Repository
public class NrtRunDefinitionRepository {

    private static final Logger log = LoggerFactory.getLogger(NrtRunDefinitionRepository.class);

    private static final String SUMMARY_COLUMNS = """
            definition_id      AS definition_id,
            name               AS name,
            description        AS description,
            scope              AS scope,
            scenario_name      AS scenario_name,
            reference_version  AS reference_version,
            target_version     AS target_version,
            ref_commandline    AS ref_commandline,
            target_commandline AS target_commandline,
            created_at         AS created_at,
            updated_at         AS updated_at,
            is_active          AS is_active
            """;

    private final NamedParameterJdbcTemplate jdbc;
    private final ObjectMapper                json;

    public NrtRunDefinitionRepository(
            @Qualifier("auditJdbcTemplate") NamedParameterJdbcTemplate jdbc,
            @Qualifier("jsonbObjectMapper") ObjectMapper json) {
        this.jdbc = jdbc;
        this.json = json;
    }

    @Retry(name = "database")
    public List<NrtRunDefinitionSummary> getAll() {
        var sql = "SELECT " + SUMMARY_COLUMNS +
                  "FROM nrt_run_definitions WHERE is_active = true ORDER BY name";
        return jdbc.query(sql, (rs, i) -> mapSummary(new NrtRunDefinitionSummary(), rs));
    }

    @Retry(name = "database")
    public NrtRunDefinitionDto getById(UUID id) {
        var sql = "SELECT " + SUMMARY_COLUMNS + ", " +
                  "reference_db::text       AS reference_db_json, " +
                  "target_db::text          AS target_db_json, " +
                  "compare_settings::text   AS compare_settings_json, " +
                  "output_settings::text    AS output_settings_json, " +
                  "column_schema::text      AS column_schema_json " +
                  "FROM nrt_run_definitions WHERE definition_id = :id";
        try {
            return jdbc.queryForObject(sql, new MapSqlParameterSource("id", id), (rs, i) -> {
                var dto = new NrtRunDefinitionDto();
                mapSummary(dto, rs);
                dto.setReference(deserialize(rs.getString("reference_db_json"), DbSettingsDto.class, DbSettingsDto::new));
                dto.setTarget(deserialize(rs.getString("target_db_json"), DbSettingsDto.class, DbSettingsDto::new));
                dto.setCompare(deserialize(rs.getString("compare_settings_json"), CompareSettingsDto.class, CompareSettingsDto::new));
                dto.setOutput(deserialize(rs.getString("output_settings_json"), OutputSettingsDto.class, OutputSettingsDto::new));
                ColumnDef[] schema = deserialize(rs.getString("column_schema_json"), ColumnDef[].class, () -> new ColumnDef[0]);
                dto.setColumnSchema(schema);
                return dto;
            });
        } catch (EmptyResultDataAccessException ex) {
            return null;
        }
    }

    @Retry(name = "database")
    public UUID create(SaveNrtRunDefinitionRequest req) {
        var sql = """
                INSERT INTO nrt_run_definitions
                    (name, description, scope, scenario_name, reference_version, target_version,
                     ref_commandline, target_commandline,
                     reference_db, target_db, compare_settings, output_settings, column_schema)
                VALUES
                    (:name, :description, :scope, :scenarioName, :referenceVersion, :targetVersion,
                     :refCommandLine, :targetCommandLine,
                     :referenceDb::jsonb, :targetDb::jsonb,
                     :compareSettings::jsonb, :outputSettings::jsonb, :columnSchema::jsonb)
                RETURNING definition_id
                """;
        return jdbc.queryForObject(sql, saveParams(null, req), UUID.class);
    }

    @Retry(name = "database")
    public boolean update(UUID id, SaveNrtRunDefinitionRequest req) {
        var sql = """
                UPDATE nrt_run_definitions
                SET name               = :name,
                    description        = :description,
                    scope              = :scope,
                    scenario_name      = :scenarioName,
                    reference_version  = :referenceVersion,
                    target_version     = :targetVersion,
                    ref_commandline    = :refCommandLine,
                    target_commandline = :targetCommandLine,
                    reference_db       = :referenceDb::jsonb,
                    target_db          = :targetDb::jsonb,
                    compare_settings   = :compareSettings::jsonb,
                    output_settings    = :outputSettings::jsonb,
                    column_schema      = :columnSchema::jsonb
                WHERE definition_id = :id AND is_active = true
                """;
        return jdbc.update(sql, saveParams(id, req)) > 0;
    }

    @Retry(name = "database")
    public boolean softDelete(UUID id) {
        var sql = "UPDATE nrt_run_definitions SET is_active = false " +
                  "WHERE definition_id = :id AND is_active = true";
        return jdbc.update(sql, new MapSqlParameterSource("id", id)) > 0;
    }

    private MapSqlParameterSource saveParams(UUID id, SaveNrtRunDefinitionRequest r) {
        var p = new MapSqlParameterSource()
                .addValue("id",                id)
                .addValue("name",              r.getName())
                .addValue("description",       r.getDescription())
                .addValue("scope",             r.getScope())
                .addValue("scenarioName",      r.getScenarioName())
                .addValue("referenceVersion",  r.getReferenceVersion())
                .addValue("targetVersion",     r.getTargetVersion())
                .addValue("refCommandLine",    r.getRefCommandLine())
                .addValue("targetCommandLine", r.getTargetCommandLine())
                .addValue("referenceDb",       serialize(r.getReference()))
                .addValue("targetDb",          serialize(r.getTarget()))
                .addValue("compareSettings",   serialize(r.getCompare()))
                .addValue("outputSettings",    serialize(r.getOutput()))
                .addValue("columnSchema",      serialize(r.getColumnSchema()));
        return p;
    }

    private static <S extends NrtRunDefinitionSummary> S mapSummary(S dto, java.sql.ResultSet rs) throws java.sql.SQLException {
        dto.setDefinitionId((UUID) rs.getObject("definition_id"));
        dto.setName(rs.getString("name"));
        dto.setDescription(rs.getString("description"));
        dto.setScope(rs.getString("scope"));
        dto.setScenarioName(rs.getString("scenario_name"));
        dto.setReferenceVersion(rs.getString("reference_version"));
        dto.setTargetVersion(rs.getString("target_version"));
        dto.setRefCommandLine(rs.getString("ref_commandline"));
        dto.setTargetCommandLine(rs.getString("target_commandline"));
        dto.setCreatedAt(toOdt(rs.getTimestamp("created_at")));
        dto.setUpdatedAt(toOdt(rs.getTimestamp("updated_at")));
        dto.setActive(rs.getBoolean("is_active"));
        return dto;
    }

    private static OffsetDateTime toOdt(Timestamp ts) {
        return ts == null ? null : ts.toInstant().atOffset(ZoneOffset.UTC);
    }

    private String serialize(Object value) {
        if (value == null) return null;
        try {
            return json.writeValueAsString(value);
        } catch (JsonProcessingException ex) {
            log.warn("JSONB serialise failure for {} — writing null", value.getClass().getSimpleName(), ex);
            return null;
        }
    }

    private <T> T deserialize(String raw, Class<T> type, java.util.function.Supplier<T> fallback) {
        if (raw == null || raw.isBlank()) return fallback.get();
        try {
            return json.readValue(raw, type);
        } catch (JsonProcessingException ex) {
            log.warn("JSONB deserialise failure for {} — using fallback", type.getSimpleName(), ex);
            return fallback.get();
        }
    }
}
